### PR TITLE
Sonarqube buildkite integration

### PR DESCRIPTION
**Pull request by pazone**
- Enhancement

## What does this PR do?
Migrates to sonar.elastic.dev and buildkite

### Pull request analysis
On pull request sonar will show the coverage for the newly added or changed codebase and run static analysis i.e. check for vulnerabilities and code smalls. See sonarqube comments below.

We can setup a quality gate by specifying the code coverage threshold. I temporarily set up 80% threshold and added an uncovered go file. So it prevents the pull request to be merged and adds a corresponding comment. 

Screenshot 2023-08-01 at 15 41 44

If no go files are changed in a PR we will see "No coverage data" message like in this comment:
Screenshot 2023-08-01 at 15 37 01

We can also configure a quality profile for any project. It determines what static checks will be performed agains our codebase. Additionally we can modify the set of rules for static analysis. Furthermore it's possible to create our own custom rules.

### Branches analysis 

When a PR merged to a target branch sonar will analyse the entire codebase. The results will be available on the [sonarqube elastic-agent dashboard](https://sonar.elastic.dev/dashboard?id=elastic_elastic-agent_AYluowg0xMq8P7b4moiZ). Currently there's no data because this PR is not yet merged.

Closes : https://github.com/elastic/elastic-agent/pull/3108

### CODE REVIEW

**Comment by pazone** [comment]
@alexsapran updated it corresponding to your comments. Thanks

**Comment by rhass** [comment]
It looks like SonarQube isn't ingesting the coverage reports because the format of the report file is not what it expects. It looks like mage is generating a junit report, and sonar can parse these if you use the `sonar.coverageReportPaths` property. The `sonar.go.tests.reportPaths` property tells Sonar that you are passing in the output file from `go test -coverprofile=example.out` for which none of the artifacts which are stored appear to have the same content. The `TEST-go-integration.out` is close, but it doesn't look like it is an exact match for the `-coverprofile` output.

I hope this helps sort things out for you folks!

**Comment by alexsapran** [comment]
Sonar will show you the coverage compared to the previous report it has collected. Given that we just added Sonar to this account, there are no previous data, thus it might look empty in the beginning.

@jlind23 we might need to adjust the sonar profile to the previous one we had, to do the code quality gate for the new code.

**Comment by jlind23** [comment]
@alexsapran sonarcloud check is required for this repo, we should probably remove it.
I'm not I sure I have access to the other sonarqube instance otherwise happy to give it a try.

**Comment by alexsapran** [comment]
oh, sorry I mixed it. Yes I removed the required check

File: .buildkite/scripts/steps/unit-tests.sh
Comment (pazone) [reply .buildkite/scripts/steps/unit-tests.sh]: Removed it from `bootstrap.sh`. However I see that we skip go installation if any version of go is installed. But go version should exactly match the `.go-version` 🤔

File: .buildkite/pipeline.yml
Comment (pazone) [reply .buildkite/pipeline.yml:22]: It actually wasn't look like an OOM. There were general uni tests failures. This image is based on Debian. Maybe that's why

File: .buildkite/scripts/steps/unit-tests.sh
Comment (alexsapran) [reply .buildkite/scripts/steps/unit-tests.sh]: I have done work on https://github.com/elastic/elastic-agent/pull/2727 to add smoother support for installing go/mage/... so I would ask to wait for my PR to get merged, hopeing to do some final testing today and then use the functions from there.

**Comment by jlind23** [comment]
@alexsapran when will sonarqube be a check? We need this one to be mandatory to avoid lack of code coverage etc..

**Comment by rhass** [comment]
> @alexsapran sonarcloud check is required for this repo, we should probably remove it. I'm not I sure I have access to the other sonarqube instance otherwise happy to give it a try.

The internal SonarQube instance is integrated at the github org level for our repos. You should be able to change the required check to use the `Elastic SonarQube` check.

image

**Comment by alexsapran** [comment]
@rhass, yes I have enabled the old check back. 

It would be good to setup the group `@elastic/ingest-eng-prod` as admins for the projects we are planning to onboard to the sonar. This way we can configure our analysis profiles and other knobs.

The removal of the Sonar and enablement of the new should be happening in one PR.

**Comment by rhass** [comment]
> @rhass, yes I have enabled the old check back.
> 
> It would be good to setup the group `@elastic/ingest-eng-prod` as admins for the projects we are planning to onboard to the sonar. This way we can configure our analysis profiles and other knobs.
> 
> The removal of the Sonar and enablement of the new should be happening in one PR.

I totally agree. We are looking into how we can leverage the SAML/Okta groups to accomplish this task as the RBAC in SonarQube 10.x has improved, but there is some automation we haven't the resources to implement right now to link up Sonar projects to the groups and assign the required permissions.

For now, I setup a snowflake group in SonarQube with @elastic/ingest-eng-pro as admins for the elastic-agent project. You should also have permissions to make new quality profiles and configure them as you see fit. The same is true for quality gates as well. Let me know if you folks have any issues. :smile:

**Comment by pazone** [comment]
@alexsapran @mrodm Let's finally review this PR. For my perspective it's sweet :)

**Comment by pazone** [comment]
@alexsapran @mrodm I think we can potentially remain using sonar cloud for public repos. Let's discuss it before we merge this PR

**Comment by pazone** [comment]
@alexsapran I've enabled 40%  coverage quality gate as it was in sonar cloud

**Comment by pazone** [comment]
buildkite build this

**Comment by alexsapran** [comment]
Looking at this quickly, I noticed that we are missing the same quality profiles on Elastic Sonar Cube as the ones we had configured in SonarCloud. 

Is this something you can do @pazone or @jlind23, Permissions where updated by @rhass, and should have access

**Comment by jlind23** [comment]
I just created a new quality gate called "Ingest Quality Gates". Should be good by now.


Screenshot 2023-09-01 at 14 58 20

**Review by alexsapran — APPROVED** [review]
code LGTM. I had left some comments around the quality profiles, but those can be changed regardless of this.

**Comment by pazone** [comment]
@jlind23 I though we are going to have separate quality gates for different projects. For example elastic-agent needs 40% coverage and another project can have a different value.

**Review by pchila — APPROVED** [review]
Change looks reasonable, however there's not much info on the dashboard for this branch https://sonar.elastic.dev/dashboard?id=elastic_elastic-agent_AYluowg0xMq8P7b4moiZ&pullRequest=3108

I suppose it's because there are no run for main :thinking: 

Would it be possible to have a manual run on main before merging this so we can have a proper look ?

**Comment by jlind23** [comment]
@pazone we can indeed have different quality gates per project, I just created one to be unblocked here.

**Comment by pazone** [comment]
@pchila That's right. We run the pipeline.yml on main branch when a PR is merged there. However I see that we also run integration tests on main branch. Correct me if I'm wrong but for now it's enough to run unit tests and publish coverage results to sonarqube

**Pull request merged**
